src/awkward/_v2/forms/numpyform.py

```diff
@@ -0,0 +1,147 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+from __future__ import absolute_import
+
+try:
+    from collections.abc import Iterable
+except ImportError:
+    from collections import Iterable
+
+import copy
+
+from awkward._v2.forms.form import Form
+
+import awkward as ak
+
+np = ak.nplike.NumpyMetadata.instance()
+
+_primitive_to_dtype = {
+    "bool": np.dtype(np.bool_),
+    "int8": np.dtype(np.int8),
+    "uint8": np.dtype(np.uint8),
+    "int16": np.dtype(np.int16),
+    "uint16": np.dtype(np.uint16),
+    "int32": np.dtype(np.int32),
+    "uint32": np.dtype(np.uint32),
+    "int64": np.dtype(np.int64),
+    "uint64": np.dtype(np.uint64),
+    "float32": np.dtype(np.float32),
+    "float64": np.dtype(np.float64),
+    "complex64": np.dtype(np.complex64),
+    "complex128": np.dtype(np.complex128),
+    "datetime64": np.dtype(np.datetime64),
+    "timedelta64": np.dtype(np.timedelta64),
+}
+
+if hasattr(np, "float16"):
+    _primitive_to_dtype["float16"] = np.dtype(np.float16)
+if hasattr(np, "float128"):
+    _primitive_to_dtype["float128"] = np.dtype(np.float128)
+if hasattr(np, "complex256"):
+    _primitive_to_dtype["complex256"] = np.dtype(np.complex256)
+
+_dtype_to_primitive = {}
+for primitive, dtype in _primitive_to_dtype.items():
+    _dtype_to_primitive[dtype] = primitive
+
+
+def from_dtype(dtype, parameters={}, inner_shape=[]):
+    if "[" in str(dtype):
+        dtype = str(dtype).split("[")
+        # WIP
+        params = copy.deepcopy(parameters)
+        params["__unit__"] = dtype[1][:-1]
+        dtype = dtype[0]
+        return NumpyForm(dtype, parameters=params, inner_shape=inner_shape)
+    else:
+        return NumpyForm(
+            _dtype_to_primitive[dtype], parameters=parameters, inner_shape=inner_shape
+        )
+
+
+def from_iter(input):
+    if isinstance(input, str):
+        return NumpyForm(primitive=input)
+    else:
+        primitive = input["primitive"]
+        inner_shape = input["inner_shape"] if "inner_shape" in input else []
+        has_identities = input["has_identities"] if "has_identities" in input else False
+        parameters = input["parameters"] if "parameters" in input else {}
+        form_key = input["form_key"] if "form_key" in input else None
+        return NumpyForm(primitive, inner_shape, has_identities, parameters, form_key)
```

Comment:
This rule should be in the generic `from_iter`; maybe I used the wrong fully qualified path in the tests, but there should only be one `from_iter` function with all the rules.